object/object.go

```diff
@@ -121,3 +129,83 @@ func SplitSlabs(slabs []Slab, lengths []int) [][]SlabSlice {
 func SingleSlabs(slabs []Slab, length int) []SlabSlice {
 	return SplitSlabs(slabs, []int{length})[0]
 }
+
+type streamReader struct {
+	key []byte
+	c   *chacha20.Cipher
+	r   io.Reader
+
+	counter uint64 // bytes, not blocks
+	nonce   uint64
+}
+
+func (sr *streamReader) Read(p []byte) (int, error) {
+	n, err := sr.r.Read(p)
+	sr.counter += uint64(n)
+	if sr.counter < 64*math.MaxUint32 {
+		sr.c.XORKeyStream(p[:n], p[:n])
+		return n, err
+	}
+	// counter overflow; xor remaining bytes, then increment nonce and xor again
+	rem := 64*math.MaxUint32 - (sr.counter - uint64(n))
+	sr.counter -= 64 * math.MaxUint32
+	sr.c.XORKeyStream(p[:rem], p[:rem])
+	// NOTE: we increment the last 8 bytes because XChaCha uses the
+	// first 16 bytes to derive a new key; leaving them alone means
+	// the key will be stable, which might be useful.
+	sr.nonce++
+	nonce := make([]byte, 24)
+	binary.LittleEndian.PutUint64(nonce[16:], sr.nonce)
+	sr.c, _ = chacha20.NewUnauthenticatedCipher(sr.key, nonce)
+	sr.c.XORKeyStream(p[rem:], p[rem:])
+	return n, err
+}
+
+type streamWriter struct {
+	key []byte
+	c   *chacha20.Cipher
+	w   io.Writer
+
+	counter uint64
+	nonce   uint64
+}
+
+func (sw *streamWriter) Write(src []byte) (n int, err error) {
+	c := make([]byte, len(src))
+	if sw.counter+uint64(len(src)) < 64*math.MaxUint32 {
+		sw.c.XORKeyStream(c, src)
+		n, err = sw.w.Write(c)
+		sw.counter += uint64(n)
+		if n != len(src) && err == nil { // should never happen
+			err = io.ErrShortWrite
+		}
+		return
+	}
+	// counter overflow, xor remaining bytes, then increment nonce and xor again
+	rem := 64*math.MaxUint32 - sw.counter
+	sw.counter -= 64 * math.MaxUint32
+	sw.c.XORKeyStream(c[:rem], src[:rem])
+	// NOTE: we increment the last 8 bytes because XChaCha uses the
+	// first 16 bytes to derive a new key; leaving them alone means
+	// the key will be stable, which might be useful.
+	sw.nonce++
+	nonce := make([]byte, 24)
+	binary.LittleEndian.PutUint64(nonce[16:], sw.nonce)
+	sw.c, _ = chacha20.NewUnauthenticatedCipher(sw.key, nonce)
+	sw.c.XORKeyStream(c[rem:], src[rem:])
+	n, err = sw.w.Write(c)
+	sw.counter += uint64(n)
+	if n != len(src) && err == nil { // should never happen
+		err = io.ErrShortWrite
+	}
+	return
```

Comment:
guess we should factor this out:

```go

type rekeyStream struct {
	key []byte
	c   *chacha20.Cipher

	counter uint64
	nonce   uint64
}

func (rs *rekeyStream) XORKeyStream(dst, src []byte) {
	rs.counter += uint64(len(src))
	if rs.counter < 64*math.MaxUint32 {
		rs.c.XORKeyStream(dst, src)
		return
	}
	// counter overflow; xor remaining bytes, then increment nonce and xor again
	rem := 64*math.MaxUint32 - (rs.counter - uint64(len(src)))
	rs.counter -= 64 * math.MaxUint32
	rs.c.XORKeyStream(dst[:rem], src[:rem])
	// NOTE: we increment the last 8 bytes because XChaCha uses the
	// first 16 bytes to derive a new key; leaving them alone means
	// the key will be stable, which might be useful.
	rs.nonce++
	nonce := make([]byte, 24)
	binary.LittleEndian.PutUint64(nonce[16:], rs.nonce)
	rs.c, _ = chacha20.NewUnauthenticatedCipher(rs.key, nonce)
	rs.c.XORKeyStream(dst[rem:], src[rem:])
}
```

then you can go back to using `cipher.StreamReader`/`cipher.StreamWriter`.